### ISSUE TITLE
Licensing issue regarding to the FDK-AAC library

### ISSUE THREAD

**Issue by Lin-Buo-Ren** [snapcrafters/ffmpeg#32]
The licensing of FDK-AAC seems to be controversial (thus the --enable-nonfree flag), we probably should evaluate whether we should really include it in the snap.

TLDR:

* Not compatible with GPL, thus conflict with the `--enable-gpl` flag

References:

* [Various Licenses and Comments about Them - GNU Project - Free Software Foundation](https://www.gnu.org/licenses/license-list.html#fdk)

**Comment by Lin-Buo-Ren** [comment]
We may want to consider making another `ffmpeg-nonfree` snap that isn't distributed on the Snap Store but allow users to build from source locally.

**Comment by Lin-Buo-Ren** [comment]
There two concerns that may affect the legitimacy of hosting the snap on Snap Store:

* The hosted snap contains code that is not redistributable by the license
* The hosted snap contains a combination of code that infringes other right holder's copyright

The inclusion of FDK-AAC may be considered to be a copyright infringement to the linked GPL components(e.g. x264/x265) as the additional limitation in FDK-AAC's license is allowed in GPL

**Comment by Lin-Buo-Ren** [comment]
Also, there are non-redistributable components in FFmpeg, the only way to allow users to access them via snap is to build from source locally.

**Comment by llogan** [comment]
Do not distribute `ffmpeg` configured with `--enable-nonfree`. The resulting license is non-free and **non-redistributable**.

**Comment by veikk0** [comment]
I think the problem here is that the maintainer(s) of this Snap have misunderstood what the FFmpeg project refers to by "non-free" in the context of compiling and distributing their software. Some of these misunderstanding should be straightened out by reading the [LICENSE](http://git.videolan.org/?p=ffmpeg.git;a=blob;f=LICENSE.md;h=613070e1b631a919a74c862f2a75a0f249a631ff;hb=HEAD) file of the project.

TL;DR: Building FFmpeg with `--enable-nonfree` makes the resulting binary potentially unredistributable, because the licenses of FFmpeg and the various external libraries can conflict with each other, even if they're all FOSS licenses. "Potentially", because disclaimer, you should always ask a lawyer, this is not legal advice, etc.. But in this case it's pretty clear, since every legal entity I know that distributes FFmpeg binaries (major Linux distros, mainly) doesn't build it with `--enable-nonfree`.

The basic license structure of FFmpeg is as follows:

* The "base" of FFmpeg is licensed under LGPLv2.1, with some files having various "MIT/X11/BSD-style licenses". The latter license are very permissive and allow re-licensing, and thus when combined the LGPLv2.1 applies. There is a clear license and the result is redistributable.
* Some optional parts are licensed under GPLv2+ (the "version 2 or later" clause) and can be enabled with the `--enable-gpl` build flag. Since the LGPLv2.1 and GPLv2 [are compatible](https://en.wikipedia.org/wiki/GNU_General_Public_License#Compatibility_and_multi-licensing), the resulting binary is under GPLv2 and is redistributable.
* Some other parts are licensed under GPLv3 and can be enabled with `--enable-version3`. Since LGPLv2, GPLv2+, and GPLv3 are compatible licenses, the resulting binary is GPLv3 and redistributable.
* There are many external libraries, and depending on their individual licenses, a library can either be compatible or incompatible with the licenses used by FFmpeg and the specific build of FFmpeg you're producing. GPLv2+ libraries are fine as long as you pass `--enable-gpl`, and GPLv3 is also fine if you pass `--enable-version3`. **Including libraries with licenses that are not compatible with LGPLv2.1, GPLv2+, or GPLv3 will result in a binary with no valid license.** Such binaries are perfectly okay to generate and use, and is done with the `--enable-nonfree` flag. However, these binaries can't be legally redistributed, since you're now combining code with conflicting license requirements, and redistributing the result would infringe on the copyright of anyone with code in either component with the conflicting licenses. Fraunhofer FDK AAC's license is considered to be incompatible with any version of the GPL [by the FSF](https://www.gnu.org/licenses/license-list.html#fdk) and falls under the "can't be redistributed in FFmpeg bulids" umbrella.

As a side note, the last point is why FFmpeg compilation guides, scripts, and build helper tools like [this](https://github.com/rdp/ffmpeg-windows-build-helpers) or [this](https://github.com/m-ab-s/media-autobuild_suite) are so prevalent, and why no Linux distro I know of distributes FFmpeg binaries build with libfdk-aac, not even in Multiverse-like repositories. Some years ago, when made aware of the issue, [HandBrake also stopped including libfdk-aac](https://forum.handbrake.fr/viewtopic.php?f=33&t=34143) but retained the option to include it at build-time.

### Solutions

As far as I've been able to determine, there are only two possible ways to resolve this.

1. Build the snap with LGPLv2.1 components only, ie. omit the `--enable-gpl` and `--enable-version3` build flags and any flags for external libraries with non-LGPL licenses. According to FFmpeg's LICENSE file, to the best of the project's knowledge, the FDK AAC license is compatible with LGPLv2.1, so only using those components with this license would result in a redistributable binary while still including libfdk. This wouldn't be ideal, though, since widely used libraries like x264 and x265 are GPLv2.
2. Don't include libfdk, ie. omit `--enable-libfdk-aac` (and `--enable-nonfree`, since there shouldn't be a reason to use this for anything intended for redistribution anyway).

In relation to this, it seems like [the OBS snap](https://github.com/snapcrafters/obs-studio) also includes a build of FFmpeg with libfdk-aac and conflicting licenses. If necessary, I'll open an issue in that repo too.

**Comment by Lin-Buo-Ren** [comment]
We probably should ship a non-distributable ffmpeg snap in the source recipe form, just so the user can choose to build for their own usage.

**Comment by veikk0** [comment]
Looks like this wasn't resolved in master, only the ffmpeg-4.3 branch. The Snap is still built with `--enable-libfdk-aac` and `--enable-nonfree`. Furthermore, 3b3369657988e275edc36ff95e4533ee967f1e4b actually bumped the libfdk-aac version.
